### PR TITLE
PA Convection Operator [convection-pa]

### PR DESCRIPTION
**Pull request by artv3**
This PR adds PA convection. 

| PR | Author | Editor | Reviewers  | Assignment | Approval | Merge |
| --- | --- | --- | ---  | --- | --- | --- |
| https://github.com/mfem/mfem/pull/914 | @artv3| @v-dobrev | @vladotomov + @pazner | 12/19/19 | 01/24/20 | ⌛due 01/30/20 |

### CODE REVIEW

**Comment by jdahm** [comment]
This is a great start, but technically there are a lot of inconsistencies that make it an apples to oranges comparison. The two codes use different devices, different threading modes, and different algorithms - basically, they are completely different codes that happen to get the same answer.

To make this consistent and useful comparison, move the matrix to the device and use cusparse or something for the matrix-vector products. The alternative is to do the PA actions in serial on the CPU, but that isn't a realistic use-case.

If it's correctly implemented, I predict FA should be better until around order 3, then PA and MF will be better, assuming you roughly fill up the device (~100k-1M DOF).

Other aspects to add to the comparison:
1. Comparing assembly times: At very low order, the full assembly will be faster since it fuses the scatter/gather, but the assembly time might kill the benefit, if not correctly implemented.
2. What is the effect on (1) when using various preconditioners (again, those need to be implemented on the GPU)?
3. Once the assembly is a fair comparison (1), then yes, test the realistic case of nonlinear advection case where the operator is re-assembled after every time-step.

Again, this is great work - we just need to be very careful of the comparisons made in order not to prematurely draw incorrect conclusions.

**Comment by artv3** [comment]
Thanks @jdahm . I agree the fair comparison would be to have both running on the same threading model. In this case serial. Even comparing a serial implementations (i.e. Device::Configure("cpu") ) I'm noticing that the PA version is much faster. Thus I'm wondering if the full assembly version is doing extra work.

File: fem/bilininteg_ext.cpp [hunk 1]
```diff
@@ -753,10 +1018,16 @@ void MassIntegrator::Assemble(const FiniteElementSpace &fes)
...
-   vec.SetSize(ne*nq);
+
+  static bool Mass_SetSize = true;
+  if(Mass_SetSize) {
+    quad1D = IntRules.Get(Geometry::SEGMENT, ir->GetOrder()).GetNPoints();
+    geom = GeometryExtension::Get(fes,*ir);
+    maps = DofToQuad::Get(fes, fes, *ir);
+    vec.SetSize(ne*nq);
+    Mass_SetSize = false;
+  }
```
Comment (artv3) [inline fem/bilininteg_ext.cpp]: If we have to reassemble operators at every right hand side evaluation it might be good to fetch this data once, would this be fair? Similar case for other operators.

**Comment by rcarson3** [comment]
Depending on what system you're running these on, you could also look into using one of the newer versions of Score-P to get some basic profiling info out to see what specific functions in each is responsible for a majority of the computation time. I've used it on one of the development machines at LLNL for Sierra with relative ease for code run on just the CPU. Although the most recent version of Score-p available on that machine, I found does have some issues with demangling the function names, but I've still found even the mangled names readable enough for my purposes with ExaConstit. I pretty much just followed ORNL's guide for using it: [https://www.olcf.ornl.gov/software_package/score-p/](https://www.olcf.ornl.gov/software_package/score-p/).

File: fem/fespace.cpp [hunk 2]
```diff
@@ -791,7 +791,7 @@ const Operator *FiniteElementSpace::GetElementRestriction(
 {
    // Check if we have a discontinuous space using the FE collection:
    const L2_FECollection *dg_space = dynamic_cast<const L2_FECollection*>(fec);
-   if (dg_space) { return NULL; }
+   //if (dg_space) { return NULL; }
```
Comment (artv3) [inline fem/fespace.cpp]: @v-dobrev , why do we have to return NULL here? I was able to run my continuous galerkin advection solver by commenting the return statement.

Comment (v-dobrev) [reply fem/fespace.cpp]: This is a simple optimization -- if the element restriction is the identity, there is no need to apply it.

Comment (artv3) [reply fem/fespace.cpp]: If I remove the comment from the check, then this example segfaults. I suspect its because I'm not using DG here, I was hoping to impose periodicity in the domain by using one of the periodic meshes. Might there be a better way to impose periodicity?

Comment (v-dobrev) [reply fem/fespace.cpp]: Oh, I see -- the issue is that periodic meshes use DG space to represent the geometry and that case is not handled in the construction of the `GeometricFactors`. This should be simple to fix in the constructor of class `GeometricFactors`.

Comment (artv3) [reply fem/fespace.cpp]: Let me see if I can cook something up for the fix discussed above. I'll also start the MPI version on this PR.

File: mesh/mesh.cpp [hunk 3]
```diff
@@ -9577,6 +9577,9 @@ GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
    const int NQ   = ir.GetNPoints();
 
    Vector Enodes(vdim*ND*NE);
+
+   //Support for DG space missing
+
```
Comment (artv3) [inline mesh/mesh.cpp]: @v-dobrev , in the case for DG elem_restr is NULL. I was wondering how to handle for that case here?

**Comment by artv3** [comment]
Hi folks, I was wondering if I could start getting some eyes on this PR.

File: fem/bilininteg.cpp [hunk 4]
```diff
@@ -822,6 +822,28 @@ void BoundaryMassIntegrator::AssembleFaceMatrix(
...
+                                                     ElementTransformation &Trans)
+{
+   int order;
+   if (trial_fe.Space() == FunctionSpace::Pk)
+   {
+      order = trial_fe.GetOrder() + test_fe.GetOrder() - 2;
+   }
+   else
+   {
+      // order = 2*el.GetOrder() - 2;  // <-- this seems to work fine too
```
Comment (v-dobrev) [inline fem/bilininteg.cpp]: This comment was copy-pasted here, right? I don't think it applies here, so it should be removed.

File: fem/fespace.cpp [hunk 5]
```diff
@@ -791,8 +791,14 @@ const Operator *FiniteElementSpace::GetElementRestriction(
 {
    // Check if we have a discontinuous space using the FE collection:
    const L2_FECollection *dg_space = dynamic_cast<const L2_FECollection*>(fec);
-   if (dg_space) { return NULL; }
-   // TODO: support other DG collections.
```
Comment (v-dobrev) [inline fem/fespace.cpp]: I don't think this TODO was addressed here, so this comment should not be removed.

File: mesh/mesh.cpp [hunk 6]
```diff
@@ -9674,6 +9674,9 @@ GeometricFactors::GeometricFactors(const Mesh *mesh, const IntegrationRule &ir,
...
```
Comment (v-dobrev) [inline mesh/mesh.cpp]: Unless I'm mistaken, I think this was addressed already by #1156, so there's no need to add this comment.

File: fem/bilininteg_convection.cpp [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+
+        v(0,q,e) = w_coeff*(cx*A00 + cy*A01 + cz*A02);
+        v(1,q,e) = w_coeff*(cx*A10 + cy*A11 + cz*A12);
+        v(2,q,e) = w_coeff*(cx*A20 + cy*A21 + cz*A22);
+      }
+
+    });
+
+   }//dim = 3
+   mesh->DeleteGeometricFactors();
```
Comment (v-dobrev) [inline fem/bilininteg_convection.cpp]: The geometric factors should not be deleted by integrators -- we want different integrators to be able to reuse them.

This call should be done only when the mesh is modified in a way that invalidates the geometric factors, e.g. mesh refinement, or updated node positions.

**Comment by v-dobrev** [comment]
Regarding the new examples, `cg-ex9.cpp` and `pa-ex9.cpp`: I think we should try to incorporate them in `ex9.cpp`.

Also, we should update `ex9p.cpp` to take advantage of the added support for PA and GPUs.

**Comment by vladotomov** [comment]
@artv3  please incorporate your changes to `ex9.cpp` and `ex9p.cpp` as Veselin pointed above. You can do it only for the volumetric integrals, or add a CG option to the example.

**Comment by vladotomov** [comment]
@artv3 I'd like to see a unit test that compares the action to the FA action, for different FE orders and meshes (2D, 3D), including AMR meshes.

**Comment by artv3** [comment]
Thanks @v-dobrev @vladotomov, I will start addresing the comments and should have a cleaner PR by early next week.

File: fem/fespace.cpp (hunk 2 above)
Comment (artv3) [reply fem/fespace.cpp]: This was fixed recently

File: fem/bilininteg_convection.cpp (hunk 7 above)
Comment (artv3) [reply fem/bilininteg_convection.cpp]: 👍

File: mesh/mesh.cpp (hunk 6 above)
Comment (artv3) [reply mesh/mesh.cpp]: 👍

File: fem/bilininteg.cpp (hunk 4 above)
Comment (artv3) [reply fem/bilininteg.cpp]: 👍

File: tests/unit/fem/test_conv.cpp [hunk 8]
```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2010, Lawrence Livermore National Security, LLC. Produced at
```
Comment (vladotomov) [inline tests/unit/fem/test_conv.cpp]: Great, thanks! Do we want to add this to test_pa_kernels.cpp? The name of that test seems generic enough :)

Comment (vladotomov) [reply tests/unit/fem/test_conv.cpp]: Sorry, test_pa_kernels.cpp is not in `master` yet. I'm referring to the test in `vecmass-vecdiff-dev`.

File: examples/ex9.cpp [hunk 9]
```diff
@@ -169,44 +180,65 @@ int main(int argc, char *argv[])
...
-   FiniteElementSpace fes(&mesh, &fec);
+   FiniteElementCollection *fec;
+   if(fem_type == cg) {
+     fec = new H1_FECollection(order, dim);
+   }else
+   {
+     fec = new DG_FECollection(order, dim);
+   }
+
+   FiniteElementSpace *fes = new FiniteElementSpace(&mesh, fec);
```
Comment (vladotomov) [inline examples/ex9.cpp]: Do we need fes to be a pointer?

Comment (artv3) [reply examples/ex9.cpp]: That's a good point, its not needed. I can change that.

File: examples/ex9.cpp [hunk 10]
```diff
@@ -53,6 +53,9 @@ double inflow_function(const Vector &x);
 // Mesh bounding box
 Vector bb_min, bb_max;
 
+// Fem discrization
+// (continous or discontinous galerkin)
+enum fem_disc {cg, dg};
```
Comment (vladotomov) [inline examples/ex9.cpp]: Probably capitalize?

File: tests/unit/fem/test_conv.cpp (hunk 8 above)
Comment (artv3) [reply tests/unit/fem/test_conv.cpp]: Sure we can add it into test_pa_kernels.cpp, https://github.com/mfem/mfem/pull/1167  will be merged into master soon right? I can shuffle the unit test around when it is in.

File: examples/ex9p.cpp [hunk 11]
```diff
@@ -52,28 +52,35 @@ double inflow_function(const Vector &x);
...
 private:
-   HypreParMatrix &M, &K;
-   const Vector &b;
+   ParBilinearForm &M, &K;
+   ParLinearForm &b;
+   HypreParMatrix *M_mat, *K_mat;
+   HypreParVector *B_vec;
    HypreSmoother M_prec;
    CGSolver M_solver;
+   fem_disc fem_type;
```
Comment (vladotomov) [inline examples/ex9p.cpp]: Seems that fem_type is not used in the class.

Comment (artv3) [reply examples/ex9p.cpp]: fixed :+1:

File: fem/bilininteg_convection.cpp (hunk 7 above)
Comment (artv3) [reply fem/bilininteg_convection.cpp]: Is there a way to trigger the recalculation of geometric factors (for problems where the coefficients depend on time)? Unless we delete them I find that we end up with the initial ones.

**Comment by artv3** [comment]
The code seems to work fine except when we combine CUDA with full assembly. I would think RHS evals would get done on the GPU but the solution when viewed stays stationary. Ping @v-dobrev , @camierjs

**Comment by artv3** [comment]
> The code seems to work fine except when we combine CUDA with full assembly. I would think RHS evals would get done on the GPU but the solution when viewed stays stationary. Ping @v-dobrev , @camierjs

I think I figured it out, pre-conditioners are not supported on the GPU yet.

File: fem/bilininteg_convection.cpp (hunk 7 above)
Comment (v-dobrev) [reply fem/bilininteg_convection.cpp]: The geometric factors returned by the mesh do not include any coefficients -- just purely geometric information. If a coefficient is changing, just re-assemble the bilinear form.

File: fem/bilininteg_convection.cpp [hunk 12]
```diff
@@ -0,0 +1,486 @@
...
+   quad1D = maps->nqpt;
+   pa_data.SetSize(dim*ne*nq, Device::GetMemoryType());
+
+   const int NE = ne;
+   const int NQ = nq;
+   Vector e_coeff(dim);
+   coeff.SetSize(dim*ne*nq);
+   auto h_C = Reshape(coeff.HostWrite(),dim,nq, ne);
+
+   if ( Q == nullptr)
```
Comment (vladotomov) [inline fem/bilininteg_convection.cpp]: make style

File: examples/ex9.cpp [hunk 13]
```diff
@@ -187,23 +215,41 @@ int main(int argc, char *argv[])
...
+      new TransposeIntegrator(new DGTraceIntegrator(velocity, 1.0, -0.5)));
+   }
 
-   m.Assemble();
-   m.Finalize();
-   int skip_zeros = 0;
-   k.Assemble(skip_zeros);
-   k.Finalize(skip_zeros);
-   b.Assemble();
+   if(pa && fem_type == DG_FE)
```
Comment (vladotomov) [inline examples/ex9.cpp]: Can't the PA be used for the volumetric DG integrals?

File: tests/unit/fem/test_conv.cpp [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+          Vector y(k.Size()), y_pa(k.Size());
+
+          for(int i=0; i<x.Size(); ++i) {x(i) = i/10.0;};
+
+          pak.Mult(x,y_pa);
+          k.Mult(x,y);
+
+          y_pa -= y;
+          double pa_error =- y_pa.Norml2();
+          std::cout << "  order: " << order
```
Comment (vladotomov) [inline tests/unit/fem/test_conv.cpp]: Mention `ConvectionIntegrator`, the `dim`, and conforming or nonconforming mesh in this output.

**Comment by vladotomov** [comment]
@artv3 after all commits since Wednesday (including the merge of master), you unit test is failing with big errors.

**Comment by vladotomov** [comment]
@v-dobrev we should extend the deadline to next week because some issue appeared.

**Comment by v-dobrev** [comment]
> @v-dobrev we should extend the deadline to next week because some issue appeared.

I extended the due dates by 1 week.

**Comment by vladotomov** [comment]
- [ ] make style
- [ ] Abort in ex9/ex9p when `-d cuda` and `(-fem` is not 1 or `-pa` is off)

**Comment by tzanio** [comment]
Extended due date to give more time for review.

**Comment by tzanio** [comment]
Not part of the new features in this PR, but can you address the Appveyor errors (in the `test_pa_coeff` test which is now run with CMake).

File: examples/ex9.cpp [hunk 15]
```diff
@@ -127,6 +134,12 @@ int main(int argc, char *argv[])
                   "Use binary (Sidre) or ascii format for VisIt data files.");
    args.AddOption(&vis_steps, "-vs", "--visualization-steps",
                   "Visualize every n-th timestep.");
+   args.AddOption((int*) &fem_type, "-fem", "--fem_type",
+                  "Choose fem discretization (CG or DG).");
```
Comment (pazner) [inline examples/ex9.cpp]: It is a bit confusing for the user here that DG = 0, CG = 1. Making this more explicit could help.

File: examples/ex9.cpp [hunk 16]
```diff
@@ -53,24 +53,28 @@ double inflow_function(const Vector &x);
 // Mesh bounding box
 Vector bb_min, bb_max;
 
+// Fem discrization
+// (continous or discontinous galerkin)
+enum fem_disc {DG_FE, CG_FE};
```
Comment (pazner) [inline examples/ex9.cpp]:
```suggestion
enum fem_disc {DG_FE = 0, CG_FE = 1};
```

File: examples/ex9.cpp [hunk 17]
```diff
@@ -135,27 +148,42 @@ int main(int argc, char *argv[])
...
+         break;
+      case 3:
+         ode_solver = new RK3SSPSolver;
+         break;
+      case 4:
+         ode_solver = new RK4Solver;
+         break;
+      case 6:
+         ode_solver = new RK6Solver;
+         break;
```
Comment (pazner) [inline examples/ex9.cpp]: These formatting changes can probably be reverted.

File: examples/ex9.cpp [hunk 18]
```diff
@@ -379,9 +453,17 @@ void velocity_function(const Vector &x, Vector &v)
...
+               v(0) = 1.0;
+               break;
+            case 2:
+               v(0) = sqrt(2./3.);
+               v(1) = sqrt(1./3.);
+               break;
+            case 3:
+               v(0) = sqrt(3./6.);
+               v(1) = sqrt(2./6.);
+               v(2) = sqrt(1./6.);
```
Comment (pazner) [inline examples/ex9.cpp]: Formatting changes here and below can probably be reverted.

File: examples/ex9p.cpp [hunk 19]
```diff
@@ -52,28 +52,35 @@ double inflow_function(const Vector &x);
 // Mesh bounding box
 Vector bb_min, bb_max;
 
+// Fem discrization
+// (continous or discontinous galerkin)
+enum fem_disc {DG_FE, CG_FE};
```
Comment (pazner) [inline examples/ex9p.cpp]:
```suggestion
enum fem_disc {DG_FE = 0, CG_FE = 1};
```

File: examples/ex9.cpp [hunk 20]
```diff
@@ -127,6 +134,12 @@ int main(int argc, char *argv[])
                   "Use binary (Sidre) or ascii format for VisIt data files.");
    args.AddOption(&vis_steps, "-vs", "--visualization-steps",
                   "Visualize every n-th timestep.");
+   args.AddOption((int*) &fem_type, "-fem", "--fem_type",
+                  "Choose fem discretization (CG or DG).");
+   args.AddOption(&pa, "-pa", "--partial-assembly", "-no-pa",
+                  "--no-partial-assembly", "Enable Partial Assembly.");
```
Comment (pazner) [inline examples/ex9.cpp]: Maybe change the help string to indicate that DG doesn't work with PA, e.g.:

```suggestion
                  "--no-partial-assembly", "Enable Partial Assembly (CG only).");
```

File: examples/ex9p.cpp [hunk 21]
```diff
@@ -135,6 +145,12 @@ int main(int argc, char *argv[])
                   "Use binary (Sidre) or ascii format for VisIt data files.");
    args.AddOption(&vis_steps, "-vs", "--visualization-steps",
                   "Visualize every n-th timestep.");
+   args.AddOption((int*) &fem_type, "-fem", "--fem_type",
+                  "Choose fem discretization (CG or DG).");
+   args.AddOption(&pa, "-pa", "--partial-assembly", "-no-pa",
+                  "--no-partial-assembly", "Enable Partial Assembly.");
+   args.AddOption(&device_config, "-d", "--device",
+                  "Device configuration string, see Device::Configure().");
```
Comment (pazner) [inline examples/ex9p.cpp]: Same comments as in the serial version.

File: examples/ex9p.cpp [hunk 22]
```diff
@@ -199,52 +220,81 @@ int main(int argc, char *argv[])
...
+   switch (fem_type)
+   {
+      case 0: fec = new DG_FECollection(order, dim); break;
+      case 1: fec = new H1_FECollection(order, dim); break;
+      default:
+         cout << "Uknown finite element space: "<<fem_type <<'\n';
+         return 3;
+   }
+
+   ParFiniteElementSpace fes(pmesh, fec);
```
Comment (pazner) [inline examples/ex9p.cpp]: I like that `fes` is no longer a pointer here!

**Comment by tzanio** [comment]
> Generally speaking, I think DG and CG are sufficiently different that it's good to have two separate examples: that's why DG diffusion is ex14, and not simply a `-dg` flag for ex1. It can make the example hard to follow, because it results in a lot of branching throughout the code. ex9 is the simplest DG example we have and the first example we refer people to when they want to learn about DG in MFEM, so I think it will serve us well to keep it simple and easy to understand.
> 
> Additionally, CG for advection without additional stabilization is not a great method. It's super oscillatory, and generally not the "right" thing to do, so I'm not sure that we want an example highlighting it.

I agree with this. It is totally fine if examples are similar (and in fact diff-ing them is one way to learn).

**Comment by vladotomov** [comment]
I also agree. I don't think ex9+CG is needed at all, especially since it doesn't really work without viscosity terms. The unit tests might be sufficient.

**Comment by artv3** [comment]
Thanks all, I'll give it a clean up pass today. Sounds like the changes to the examples are not needed, I'll revert them to what is in master.

**Comment by artv3** [comment]
Hi all, all the travis builds pass except for branch-history. I get:
```
=== BEGIN error log from branch-history.err

This branch adds/modifies a large file: mesh/mesh.cpp (size 267K)

Commit 9ab4d9d8 adds/modifies a large file: mesh/mesh.cpp (size 268K)

Commit 1ecafbd3 adds/modifies a large file: mesh/mesh.cpp (size 284K)

=== END error log from branch-history.er
```

 I'm not sure how to resolve that issue. Could squashing this commit fix things?

**Comment by jandrej** [comment]
I can't find 1ecafbd3 in your commit history.

**Comment by tzanio** [comment]
This is an acceptable failure, see also https://github.com/mfem/mfem/pull/1207#issuecomment-578174801 and doesn't have to be resolved.

Administrators can still merge the PR with this failure (it is just a reminder to be careful).

The `branch-history` script currently has conservative settings, we can consider relaxing them or filtering specific files...

**Comment by v-dobrev** [comment]
The changes in `mesh/mesh.cpp` done in the history of this branch are probably stored as deltas so their size is probably small. I think it is OK to merge this branch even without squashing -- after the testing in `next`.

@vladotomov, @pazner, do you want to take a look at the latest commits and confirm everything is good to go?

**Comment by pazner** [comment]
The latest commits look good to me!

File: fem/bilininteg.cpp [hunk 23]
```diff
@@ -901,6 +901,28 @@ void BoundaryMassIntegrator::AssembleFaceMatrix(
...
+   else
+   {
+      order = trial_fe.GetOrder() + test_fe.GetOrder() + trial_fe.GetDim() - 1;
+   }
+
+   if (trial_fe.Space() == FunctionSpace::rQk)
+   {
+      return RefinedIntRules.Get(trial_fe.GetGeomType(), order);
+   }
+   return IntRules.Get(trial_fe.GetGeomType(), order);
```
Comment (v-dobrev) [inline fem/bilininteg.cpp]: The order/rule returned by this method is not consistent with the one used in `ConvectionIntegrator::AssembleElementMatrix` -- I think it is better to be consistent. The right formula should be (no explicit dependence on the `FunctionSpace`):
```c++
   order = Trans.OrderGrad(&trial_fe) + Trans.Order() + test_fe.GetOrder();
```

File: tests/unit/fem/test_pa_coeff.cpp [hunk 24]
```diff
@@ -14,6 +14,8 @@
 
 using namespace mfem;
 
+const double PI = 3.14159265358979323846;
```
Comment (v-dobrev) [inline tests/unit/fem/test_pa_coeff.cpp]: If `#include "mfem.hpp"` is before `#include "catch.hpp"` then `<cmath>` will be included after the mfem config header which enables the definition of `M_PI` by `<cmath>` under windows (one needs to `#define _USE_MATH_DEFINES` before including `<cmath>` to get constants like `M_PI`): https://github.com/mfem/mfem/blob/4828e7e2a72b707f75a8e60751cffddb89c0b621/config/config.hpp#L32-L36

**Comment by vladotomov** [comment]
Looks good to me too.

File: tests/unit/fem/test_pa_coeff.cpp (hunk 24 above)
Comment (artv3) [reply tests/unit/fem/test_pa_coeff.cpp]: Ah, I was wondering why it wasn't working. Thanks! Just shuffled the code a bit.

File: fem/bilininteg.cpp (hunk 23 above)
Comment (artv3) [reply fem/bilininteg.cpp]: Yes, that is much cleaner. Fixed.

**Comment by tzanio** [comment]
May be good to highlight this in the CHANGELOG, maybe as part of #1207?

**Comment by vladotomov** [comment]
It got mentioned in commit 3ff0314, but that has gotten lost somewhere along the way :/

**Comment by v-dobrev** [comment]
It was removed in 8bf6d6b0 -- need to be more careful with merge conflicts.
